activesupport/lib/active_support/cache/streaming_compressor.rb

```diff
@@ -0,0 +1,109 @@
+# frozen_string_literal: true
+
+require "zlib"
+require "concurrent/executor/cached_thread_pool"
+require "active_support/core_ext/numeric/bytes"
+
+module ActiveSupport
+  module Cache
+    # Serializes object and compresses serialized representation in streaming way.
+    # This avoids storing whole copy of serialized intermediate representation,
+    # like this happens in `Zlib.deflate(Marshal.dump(value))`.
+    module StreamingCompressor
+      DEFLATE_BUFFER_SIZE = 16.kilobytes.to_i
+      INFLATE_CHUNK_SIZE = 1.kilobyte.to_i
+
+      extend self
+
+      # Dumps value into string. Returns nil if `:compress_threshold` option
+      # is set and serialized value is less than given value.
+      def dump(value, **options)
+        io = StringIO.new
+        return unless dump_to_io(value, io, **options)
+        io.rewind
+        io.read
+      end
+
+      # Same as dump but streams compressed data. Returns nil and don't stream
+      # anything if `:compress_threshold` option is set
+      # and serialized value is less than given value.
+      def dump_to_io(value, to_io, compress_threshold: nil)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            Marshal.dump(value, wio)
+          ensure
+            wio.close
+          end
+        end
+        if compress_threshold
+          chunk = rio.read([DEFLATE_BUFFER_SIZE, compress_threshold].max)
+          return if chunk.size < compress_threshold
+        end
+        deflate(rio, to_io, initial_chunk: chunk)
+        true
+      end
+
+      def load(value)
+        rio, wio = binary_pipe
+        within_safe_thread do
+          begin
+            inflate(value, wio)
+          ensure
+            wio.close
+          end
+        end
+        Marshal.load(rio)
+      end
+
+      def deflate(from_io, to_io, initial_chunk: nil)
+        zlib = Zlib::Deflate.new
+        to_io << zlib.deflate(initial_chunk) if initial_chunk
+        to_io << zlib.deflate(from_io.read(DEFLATE_BUFFER_SIZE)) until from_io.eof?
+        to_io << zlib.finish
+      ensure
+        zlib.close
+      end
+
+      def inflate(value, to_io)
+        zlib = Zlib::Inflate.new
+        string_to_chunks(value, INFLATE_CHUNK_SIZE).each do |chunk|
+          to_io.write(zlib.inflate(chunk))
+        end
+      ensure
+        zlib.close
+      end
+
+      private
+
+        def binary_pipe
+          rio, wio = IO.pipe(binmode: true)
+          wio.set_encoding(Encoding::BINARY)
+          [rio, wio]
+        end
+
+        def thread_pool
+          @thread_pool ||= Concurrent::CachedThreadPool.new
+        end
+
+        # Runs task within thread and rescues error.
+        def within_safe_thread
+          thread_pool.post do
+            begin
+              yield
+            rescue StandardError => e
+              handle_thread_exception(e)
+            end
+          end
+        end
+
+        def handle_thread_exception(e)
+          Rails.logger.error(e)
```

Comment:
Please advise the most appropriate solution for default exception handler in threads.